File: web/src/screens/titleScreen.ts
import { Button } from '../constants.js';
import { Screen, ScreenManager } from '../screen.js';
import { Input } from '../input.js';
import { GameState } from '../gameState.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import { Music } from '../media/music.js';
import { selectFeedback } from '../utils.js';
import type { Animation } from '../media/animation.js';
import type { BitmapFont } from '../media/font.js';
import { LevelDescription } from '../levelDescription.js';
import { LevelSelectScreen } from './levelSelectScreen.js';
import { LevelScreen } from './levelScreen.js';

export class TitleScreen extends Screen {
  private hasContinue = false;
  private readonly font: BitmapFont;
  private frameCounter = 0;
  private gameStart = false;
  private readonly hand: Animation;
  private selected = 0;
  private readonly title: Animation;

  constructor() {
    super();
    this.font = Resource.getFont('data/images/font.bmp');
    this.hand = Resource.getAnimation('data/images/hand.bmp', 1);
    this.title = Resource.getAnimation('data/images/title.bmp', 1);
  }

  override onEntered(): void {
    this.gameStart = false;
    Music.playSong('data/music/intro2.xm');
    this.hasContinue = GameState.isSavePresent();
    if (this.hasContinue) this.selected = 1;
  }

  onDraw(ctx: CanvasRenderingContext2D): void {
    this.title.drawFrame(ctx, 0, 0, 0);

    // Draw controls help text
    if (Input.hasTouch()) {
      this.font.drawCenter(ctx, 'TOUCH SCREEN TO SHOW CONTROLS', 0, 115, 320, 10);
    } else {
      this.font.drawCenter(ctx, 'ARROWS-MOVE SPACE-JUMP CTRL-ROPE', 0, 115, 320, 10);
    }

    if (!this.gameStart)
      this.hand.drawFrame(ctx, 0, 115, 150 + this.selected * 10);

    if (!this.gameStart || (this.selected !== 0 || this.frameCounter % 10 < 5))
      this.font.draw(ctx, 'NEW GAME', 126, 150);

    if (this.hasContinue) {
      if (!this.gameStart || (this.selected !== 1 || this.frameCounter % 10 < 5))
        this.font.draw(ctx, 'CONTINUE', 126, 160);
      this.font.draw(ctx, 'EXIT GAME', 126, 170);
    } else {
      this.font.draw(ctx, 'EXIT GAME', 126, 160);
    }

    this.font.draw(ctx, 'DARKBITS', 130, 215);
    this.font.draw(ctx, 'SPEEDHACK 2011', 113, 225);
  }

  onLogic(): void {
    this.frameCounter++;

    if (this.gameStart) {
      if (this.frameCounter > 100) {
        ScreenManager.add(new LevelSelectScreen());
        if (this.selected === 0) {
          ScreenManager.add(new LevelScreen(new LevelDescription(
            'tutorial', 'data/rooms/tutorial.txt', 0, 'data/music/olof9.xm')));
        }
      }
      return;
    }

    if (Input.isPressed(Button.DOWN)) {
      this.selected++;
      if (this.selected > 1 && !this.hasContinue) this.selected = 1;
      else if (this.selected > 2) this.selected = 2;
      else selectFeedback();
    }

    if (Input.isPressed(Button.UP)) {
      this.selected--;
      if (this.selected < 0) this.selected = 0;
      else selectFeedback();
    }

    if (Input.isPressed(Button.EXIT)) this.exit();

    if (Input.isPressed(Button.FIRE)) {
      if (this.selected === 0) {
        this.gameStart = true;
        this.frameCounter = 0;
        Music.stop();
        Sound.playSample('data/sounds/start');
        GameState.clear();
      }
      if (this.selected === 1 && this.hasContinue) {
        this.gameStart = true;
        this.frameCounter = 0;
        Music.stop();
        Sound.playSample('data/sounds/start');
        GameState.loadFromFile();
      }
      if (this.selected === 1 && !this.hasContinue) this.exit();
      if (this.selected === 2 && this.hasContinue) this.exit();
    }
  }
}

File: wii/src/screens/title_screen.cpp
#include "screens/title_screen.h"
#include "resource.h"
#include "media/animation.h"
#include "media/font.h"
#include "media/sound.h"
#include "media/music.h"
#include "input.h"
#include "game_state.h"
#include "constants.h"
#include "level_description.h"
#include "screens/level_select_screen.h"
#include "screens/level_screen.h"

TitleScreen::TitleScreen() {
    font_ = Resource::getFont("data/images/font.bmp");
    hand_ = Resource::getAnimation("data/images/hand.bmp", 1);
    title_ = Resource::getAnimation("data/images/title.bmp", 1);
}

void TitleScreen::onEntered() {
    gameStart_ = false;
    Music::playSong("data/music/intro2.xm");
    hasContinue_ = GameState::isSavePresent();
    if (hasContinue_) selected_ = 1;
}

void TitleScreen::onDraw(SDL_Renderer* renderer) {
    title_->drawFrame(renderer, 0, 0, 0);

    // Draw controls help text (no touch on Wii, always show keyboard/gamepad controls)
    font_->drawCenter(renderer, "ARROWS-MOVE SPACE-JUMP CTRL-ROPE", 0, 115, 320, 10);

    if (!gameStart_)
        hand_->drawFrame(renderer, 0, 115, 150 + selected_ * 10);

    if (!gameStart_ || (selected_ != 0 || frameCounter_ % 10 < 5))
        font_->draw(renderer, "NEW GAME", 126, 150);

    if (hasContinue_) {
        if (!gameStart_ || (selected_ != 1 || frameCounter_ % 10 < 5))
            font_->draw(renderer, "CONTINUE", 126, 160);
        font_->draw(renderer, "EXIT GAME", 126, 170);
    } else {
        font_->draw(renderer, "EXIT GAME", 126, 160);
    }

    font_->draw(renderer, "DARKBITS", 130, 215);
    font_->draw(renderer, "SPEEDHACK 2011", 113, 225);
}

void TitleScreen::onLogic() {
    frameCounter_++;

    if (gameStart_) {
        if (frameCounter_ > 100) {
            ScreenManager::add(new LevelSelectScreen());
            if (selected_ == 0) {
                ScreenManager::add(new LevelScreen(LevelDescription(
                    "tutorial", "data/rooms/tutorial.txt", 0, "data/music/olof9.xm")));
            }
        }
        return;
    }

    if (Input::isPressed(Button::DOWN)) {
        selected_++;
        if (selected_ > 1 && !hasContinue_) selected_ = 1;
        else if (selected_ > 2) selected_ = 2;
        else Sound::playSample("data/sounds/select");
    }

    if (Input::isPressed(Button::UP)) {
        selected_--;
        if (selected_ < 0) selected_ = 0;
        else Sound::playSample("data/sounds/select");
    }

    if (Input::isPressed(Button::EXIT)) exit();

    if (Input::isPressed(Button::FIRE)) {
        if (selected_ == 0) {
            gameStart_ = true;
            frameCounter_ = 0;
            Music::stop();
            Sound::playSample("data/sounds/start");
            GameState::clear();
        }
        if (selected_ == 1 && hasContinue_) {
            gameStart_ = true;
            frameCounter_ = 0;
            Music::stop();
            Sound::playSample("data/sounds/start");
            GameState::loadFromFile();
        }
        if (selected_ == 1 && !hasContinue_) exit();
        if (selected_ == 2 && hasContinue_) exit();
    }
}
